pkg/webhooks/pod_admission_webhook.go

```diff
@@ -0,0 +1,112 @@
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+	"sigs.k8s.io/jobset/pkg/controllers"
+	"sigs.k8s.io/jobset/pkg/util/shared"
+)
+
+//+kubebuilder:webhook:path=/validate--v1-pod,mutating=false,failurePolicy=fail,sideEffects=None,groups="",resources=pods,verbs=create,versions=v1,name=vpod.kb.io,sideEffects=None,admissionReviewVersions=v1
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (v *podWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	return v.validate(ctx, obj)
+}
+
+func (v *podWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (v *podWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (p *podWebhook) validate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("expected a Pod but got a %T", obj)
+	}
+
+	// If pod is not part of a JobSet using exclusive placement, we don't need to validate anything.
+	_, usingExclusivePlacement := pod.Annotations[jobset.ExclusiveKey]
+
+	// If pod is part of a JobSet that is using the node selector exclusive placement starategy,
+	// we don't need to validate anything.
+	_, usingNodeSelectorStrategy := pod.Annotations[jobset.NodeSelectorStrategyKey]
+
+	// Do not validate anything for leader pods, proceed with creation immediately.
+	// For follower pods, validate leader pod exists and is scheduled.
+	if usingExclusivePlacement && !usingNodeSelectorStrategy && !shared.IsLeaderPod(pod) {
+		leaderScheduled, err := p.leaderPodScheduled(ctx, pod)
```

Comment:
we also need to check that the nodeSelector was added the mutating webhook. This is to avoid the race where the leader got scheduled between the two calls.

leader not scheduled
follower mutate webhook called, node affinity not set
leader got scheduled
follower validate called <--- here we should reject to force another call to the mutating webhook to set the node affinity

Reply:
Ah I thought the validating webhook was called before the mutating webhook. Fixed!